### DISCUSSION TITLE
overlayScrollbars-nx or overlayScrollbars?

### DISCUSSION THREAD

**Discussion by benjamin-hoefer-fnt** [KingSora/OverlayScrollbars#589] [Q&A]
Hello!

I am building an abstraction layer for overlayscrollbars in our Angular application.

I am just wondering which way you would recommend?

I initialized overlayscrollbars in two different ways:

1. By using OverlayScrollbars and this: 

`OverlayScrollbars({` target: document.querySelector('#myElement') }, {});`

2. By using OverlayScrollbars of overlayscrollbars-ngx and this:

`this.osInstance.osInitialize(this.elementRef.nativeElement);`

Both ways has the same ouptut and it actually works. But what would your recommend way as the first one is the more general approach and the second one is more the Angular approach. Should I stick to the Angular approach as I am building this for an Angular application?

Kind regards
Benjamin

**Comment by KingSora** [comment]
Good day @benjamin-hoefer-fnt :)

It depends how big the abstraction is, since `overlayscrollbars-ngx` should already be the abstraction. (It has a component and directive)

If youre dealing with just an edge case where the `overlayscrollbars-ngx` package isn't providing the right tools for it but otherwise works, I would stick with it.

In case you build your own component / directive and outside of it you arent using the `overlayscrollbars-ngx` at all, I would go with your first example. Because then you can remove the package entirely.

**Reply by benjamin-hoefer-fnt** [reply]
Good morning ;) Thank you for the reply!

As I am building an own directive I think I will stick to the more general approach. It seems to be working better anyway.

Just to let you know when I used the directive of overlayscrollbars-ngx and set the 'defer' property to 'true' before initializing and print the instance of the directive class to the console afterwards the defer property turned the instance to 'null' somehow. You might could check this behaviour.

Another question:

I am struggeling right now to add my abstraction of the scrollbar to a third party overlay (tippy.js of helipopper) as it will not be recognized there. I read about the object initialization in the documenation using 'viewport':

<img width="429" alt="image" src="https://github.com/KingSora/OverlayScrollbars/assets/139326200/3c59f32a-055f-4ada-9dc5-7d984446a818">

But to be honest it not very clear for me what the difference is between the target and the viewport. How do these two keys differ and how do I use 'viewport' properly to make the scrollbar work on third party libraries?

**Reply by KingSora** [reply]
@benjamin-hoefer-fnt the `defer` property is there to defer the initialization to a point in time where the browser is idle. This means that the instance will be `null` right after the initialization because it is deferred.. I'm now assuming thats not a bug but the behavior of the property.. To illustrate the difference again here are two examples, one with and one without `defer`:

```js
// with defer=true
osInitialize(document.body);
const instance = osInstance(); // null
// .. browser takes its time to initialize in an unknown amount of seconds
const instance = osInstance(); // not null
```
```js
// with defer=false
osInitialize(document.body);
const instance = osInstance(); // not null
```

Regarding tippy.js, are you able to describe what you wanna do and maybe also post an example (on CodeSandbox / StackBlitz)? I would then take a look and provide an solution. The object initialization is there in case a different plugin builds the DOM you need and then you want OverlayScrollbars be applied to that DOM. That happens often with virutalization libraries for example. Since you can insert your own elements into the popups of tippy.js I'm not sure whether that feature is really needed in that case.

**Reply by benjamin-hoefer-fnt** [reply]
I am afraid, but I can't provide an example so easily at the moment.

I just figured out that the problem is not the overlay library but my abstraction. As soon as I use the overlay-scollbars component on the element it works. The problem is that the overlay will be created by an ng-template and the div element on which I want to put the scrollbar is inside the ng-template. When I swap the ng-template through a div element my abstraction works.

I think the initialization object which I pass just not correct but I can not find out how to do it porperly...

**Reply by KingSora** [reply]
@benjamin-hoefer-fnt maybe the source code of `overlayscrollbars-ngx` is of any help: https://github.com/KingSora/OverlayScrollbars/tree/master/packages/overlayscrollbars-ngx/src

**Reply by benjamin-hoefer-fnt** [reply]
Oh, I was looking for this but could not find it. That's exactly what I need. Thank you for this great support!

**Reply by KingSora** [reply]
youre welcome :)

**Reply by benjamin-hoefer-fnt** [reply]
It actually works! Thanks a lot! Have a great weekend!
